src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_agents.clj

```diff
@@ -45,6 +50,92 @@
 
 (declare send-flush-instance!)
 
+(schema/defn flush-reference-pool-instance!
+  [pool-context instance cleanup-fn]
+  (let [pool-state (jruby-internal/get-pool-state-container pool-context)
+        pool (jruby-internal/get-pool pool-context)
+        config (:config pool-context)
+        _ (.releaseItem pool instance)]
+    (when (not (:flush-pending @pool-state))
+      ;; trigger a flush of the Jruby pool if one hasn't already
+      ;; been started
+      (try
+        (do
+          (swap! pool-state assoc :flush-pending true)
+          (.lock pool)
+          (.clear pool)
+          (cleanup-fn instance)
+          (.terminate (:scripting-container instance))
+          (jruby-internal/create-pool-instance! pool 1 config
+                                                (partial send-flush-instance! pool-context)
+                                                (:splay-instance-flush config))
+          (log/info (i18n/trs "Finished creating JRubyInstance 1 of 1"))
+          (swap! pool-state assoc :flush-pending false)
+          (.unlock pool))
+        (catch Exception e
+         (.clear pool)
+         (jruby-internal/insert-poison-pill pool e)
+         (throw (IllegalStateException.
+                  (i18n/trs "There was a problem flushing and refilling the pool." e))))))))
+
+(schema/defn flush-reference-pool!
+  [pool-context
+   refill?
+   on-complete]
+  (let [pool-state (jruby-internal/get-pool-state-container pool-context)
+        pool (:pool (jruby-internal/get-pool-state pool-context))
+        cleanup-fn (get-in pool-context [:config :lifecycle :cleanup])]
+    (when (not (:flush-pending @pool-state))
+      (try
+        (swap! pool-state assoc :flush-pending true)
```

Comment:
This is racy since you test and set separately.

Reply:
Also it won't be set back in case of failure.

Reply:
Note to self: use `compare-and-set!` here